src/diffusers/models/transformers/sana_transformer.py

```diff
@@ -0,0 +1,448 @@
+
+        for name, module in self.named_children():
+            fn_recursive_attn_processor(name, module, processor)
+
+    def forward(
+        self,
+        hidden_states: torch.Tensor,
+        encoder_hidden_states: torch.Tensor,
+        timestep: torch.LongTensor,
+        encoder_attention_mask: Optional[torch.Tensor] = None,
+        attention_mask: Optional[torch.Tensor] = None,
+        return_dict: bool = True,
+    ) -> Union[Tuple[torch.Tensor, ...], Transformer2DModelOutput]:
+        # ensure attention_mask is a bias, and give it a singleton query_tokens dimension.
+        #   we may have done this conversion already, e.g. if we came here via UNet2DConditionModel#forward.
+        #   we can tell by counting dims; if ndim == 2: it's a mask rather than a bias.
+        # expects mask of shape:
+        #   [batch, key_tokens]
+        # adds singleton query_tokens dimension:
+        #   [batch,                    1, key_tokens]
+        # this helps to broadcast it as a bias over attention scores, which will be in one of the following shapes:
+        #   [batch,  heads, query_tokens, key_tokens] (e.g. torch sdp attn)
+        #   [batch * heads, query_tokens, key_tokens] (e.g. xformers or classic attn)
+        if attention_mask is not None and attention_mask.ndim == 2:
+            # assume that mask is expressed as:
+            #   (1 = keep,      0 = discard)
+            # convert mask into a bias that can be added to attention scores:
+            #       (keep = +0,     discard = -10000.0)
+            attention_mask = (1 - attention_mask.to(hidden_states.dtype)) * -10000.0
+            attention_mask = attention_mask.unsqueeze(1)
+
+        # convert encoder_attention_mask to a bias the same way we do for attention_mask
+        if encoder_attention_mask is not None and encoder_attention_mask.ndim == 2:
+            encoder_attention_mask = (1 - encoder_attention_mask.to(hidden_states.dtype)) * -10000.0
+            encoder_attention_mask = encoder_attention_mask.unsqueeze(1)
+
+        # 1. Input
+        batch_size, num_channels, height, width = hidden_states.shape
+        p = self.config.patch_size
+        post_patch_height, post_patch_width = height // p, width // p
+
+        hidden_states = self.patch_embed(hidden_states)
+
+        timestep, embedded_timestep = self.time_embed(
+            timestep, batch_size=batch_size, hidden_dtype=hidden_states.dtype
+        )
+
+        encoder_hidden_states = self.caption_projection(encoder_hidden_states)
+        encoder_hidden_states = encoder_hidden_states.view(batch_size, -1, hidden_states.shape[-1])
+
+        encoder_hidden_states = self.caption_norm(encoder_hidden_states)
+
+        # 2. Transformer blocks
+        use_reentrant = is_torch_version("<=", "1.11.0")
+
+        def create_block_forward(block):
+            if torch.is_grad_enabled() and self.gradient_checkpointing:
+                return lambda *inputs: torch.utils.checkpoint.checkpoint(
```

Comment:
I was going to write a internal thread about partial grad checkpointing and how this might help making it usable for all different models we support, with some numbers from CogVideoX training, but didn't find the time yet.

The intention was to support use cases like: https://github.com/huggingface/diffusers/pull/9982#issuecomment-2517826764

TLDR; It really does help achieve a tradeoff between training speed vs memory.

The idea was to allow some sort of gradient checkpointing callback that can be set by users directly, and we would default to `torch.utils.checkpoint.checkpoint(...)` if gradient checkpointing was enabled. This would allow setting gradient checkpointing conditionally with something like:

```python
# Default
def default_block_forward(block):
    if torch.is_grad_enabled() and self.gradient_checkpointing:
        return lambda *inputs: torch.utils.checkpoint.checkpoint(
            lambda *x: block(*x), *inputs, use_reentrant=use_reentrant
        )
    else:
        return block

self.block_forward = default_block_forward

# User defined custom grad forward
def custom_block_forward(block):
    if block.layer_index % 2 == 0 and torch.is_grad_enabled() and self.gradient_checkpointing:
        return lambda *inputs: torch.utils.checkpoint.checkpoint(
            lambda *x: block(*x), *inputs, use_reentrant=use_reentrant
        )
    else:
        return block

# self.block_forward = custom_block_forward

# Training/inference with grad
self.block_forward(block)(input1, input2, ...)
```

Anyway, we can debate the design later because probably not a lot of people would be interested in this so it's relatively low prio. Will change current code to how we normally do grad checkpointing

Reply:
yeah, happy to discuss & open to change but I think it's better to change everywhere all together